### PR TITLE
Configuration Improvements and Type fixes

### PR DESCRIPTION
**Pull request by dodiego**
The project doesn't extend the original fastify typings, which causes type errors when tryting to reference:
- fastify.requireAuthentication
- fastify.generateAuthToken
- all properties from parameters are recognized as mandatory when only secret is needed

This PR address those issues and adds some others improvements to the project configuration

- added default nodejs gitignore
- added type options
- extended fastify request type with auth property
- extended fastify instance with generateAuthToken and requireAuthentication
- added jsconfig to have better intellisense and type checking
- added eslint with standardjs

### CODE REVIEW

**Comment by patrickpissurno** [comment]
Hi,

Thanks for your contribution. Before I can merge your PR, I need you to undo the "added eslint with standardjs" part of it.

I can merge all the other changes, but eslint totally broke the diff and changed the coding style of the source code.

I do see the reasoning for eslint in large codebases, where many devs have direct push access to the upstream repo. But there are shortcomings with it too, and I don't see it as a good fit for a repo this size. I didn't write a "CODESTYLE.md" file, because it's just not necessary. The codebase is small, and all contributions go through me, so I can still double-check (just as I'm doing right now).

Please format you code with 4 spaces, instead of 2. And use simple quotes (') instead of double quotes("). You'll have to undo the changes eslint has done. That's why it's nice to do only one thing per commit. This way it's a lot easier to undo only one part of it.

If I may, I'd recommend making a copy of your modified source code in a separate folder, then reverting to the upstream head with git --reset-hard, then copying only what you really changed back (and ajusting the code style to match).

Take your time to do the changes. And again, thank you for your contribution. Better IDE support is really nice and important. It'll make this plugin much more usable.

**Comment by dodiego** [comment]
Hey @patrickpissurno, just updated the files without the eslint styling, it is much cleaner and easier to review it now. 

I still believe that the project can benefit from some styling enforcement, I had to disable my code formatter because it doesn't depend on eslint rules and other developers might encounter the same problem when contributing to the project, but I still agree with you that this PR should not do that.

I can make another PR with some eslint/prettier configuration respecting the current code styling (single quotes, semicolons, trailing commas, etc), that way we keep the code as it is and anyone who contributes will not have to worry with that.

Also, great work with this project!

File: example.js
Comment (patrickpissurno) [inline example.js:8]: This should not be async. Please remove the `async` keyword.

**Reason:**

When using async request handlers, one should not use `reply.send`. Instead, they should `return` the value.

So, to keep the example short and simple, I opted not to use `async` here. Instead, relying on the callback based approach, which can be written as the one-liner that you see.

File: example.js
Comment (patrickpissurno) [inline example.js:28]: This should not be async. Please remove the `async` keyword.

The reason is the same as the above request.

**Review by patrickpissurno — CHANGES_REQUESTED** [review]
Nice!

One last thing before merging, though. There were some changes in the example.js file that shouldn't have been made. I explained why in the review.

I'm sorry for the rather long process to approve those changes. I'd fix them myself if I could, just not to bother you with that. But I'd like you to keep the authorship.

Thanks!

File: example.js
Comment (dodiego) [reply example.js:8]: So, this should be either be `async` or not have a return value.

**Reason:**

The fastify typings specify that your requests should either return a promise or use `res.send` with no return at all. Arrow functions implicitly return `res.send` which returns the `res` object. It works at runtime but breaks the type rules.

image

Also, no need for sorry at all man, PRs give us a great space for debating, if we don't agree on something we should speak our minds and improve the solution together.

File: example.js
Comment (dodiego) [reply example.js:28]: Should be either `async` or not have a return value, the reason is explained above.

**Comment by patrickpissurno** [comment]
You're right.

I'll be merging this now.

Thanks!

**Comment by patrickpissurno** [comment]
This feature is already out on NPM version 1.1.0 of fastify-esso!
`npm i --save fastify-esso@latest`

**Pull request merged**
